File: 3. Longest Substring Without Repeating Characters/solution.cpp
/*
  Runtime: 12 ms, faster than 97.64% of C++ online submissions for Longest Substring Without Repeating Characters.
  Memory Usage: 11.4 MB, less than 69.44% of C++ online submissions for Longest Substring Without Repeating Characters.
*/

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
      size_t ans = 0, sub_len = 0, j = 0;
      unordered_map<char, size_t> m;

      for(size_t i = 0; i < s.size(); i++){
          if(m.count(s.at(i)) == 1){
              if (m.at(s[i]) >= j)
              {
                  size_t pos = m.at(s.at(i));   
                  sub_len = sub_len - (pos - j) - 1;
                  j = i - sub_len;
              }
          }
          m[s.at(i)] = i;
          sub_len++;
          ans = max(ans, sub_len);
      }
      return ans;
    }
};

File: 3. Longest Substring Without Repeating Characters/solution.py
# 1
# Runtime: 64 ms, faster than 92.31% of Python3 online submissions for Longest Substring Without Repeating Characters.
# Memory Usage: 13.3 MB, less than 49.91% of Python3 online submissions for Longest Substring Without Repeating Characters.


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        ans, sub_len, sub_start, dic = 0, 0, 0, {}
        for i, c in enumerate(s):
            if c in dic and dic[c] >= sub_start:
                pos = dic[c]
                sub_len = sub_len - (pos - sub_start) - 1
                sub_start = i - sub_len
            dic[c] = i
            sub_len += 1
            ans = max(ans, sub_len)
        return ans

# 2
# Runtime: 48 ms, faster than 99.93% of Python3 online submissions for Longest Substring Without Repeating Characters.
# Memory Usage: 13.4 MB, less than 13.13% of Python3 online submissions for Longest Substring Without Repeating Characters.

class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        collector, ans, sub_start= {}, 0, 0
        for i, c in enumerate(s):
            if c in collector and collector[c] >= sub_start:
                sub_start = collector[c] + 1
            else:
                ans = max(ans, i - sub_start + 1)
            collector[c] = i
        return ans
